notebooks/working/3.0-mab-independent-variable-simulation.md

```diff
@@ -34,487 +47,269 @@ import random
+    for alt in data_long[alt_id_col].unique():
+        # Store data for specific alternative (mode)
+        alt_data = data_long.loc[data_long[alt_id_col] == alt]
+        alt_spec_dic = defaultdict(dict)
+        # Loop around the alternative specific
+        # variables in the input dictionary
+        for alt_var in alt_spec:
+            var_type = var_types[alt_var]
+            # If data is categorical
+            if isCategorical(var_type):
+                # If only one category
+                if isUnique(alt_data[alt_var]):
+                    # Add name of alternative to variable
+                    # and store distriburion & parameters
+                    alt_spec_var_dic = getConstantDistAlt(alt_var, alt_data, alt_name_dic, alt)
+                else:
+                    # If more than one category, compute
+                    # the frequency of values and store
+                    # as parameters
+                    # Add name of alternative to variable
+                    # and store distriburion & parameters
+                    alt_spec_var_dic = getCategoricalDistAlt(alt_var, alt_data, alt_name_dic, alt)
+            else:
+                # If data is not categorical but has one unique value
+                if isUnique(alt_data[alt_var]):
+                    alt_spec_var_dic = getConstantDistAlt(alt_var, alt_data, alt_name_dic, alt)
+                # If data is not categorical but has more than one unique value
+                else:
+                    alt_spec_var_dic = getContinuousDistAlt(alt_var, alt_data, alt_name_dic, alt, cont_dists)
+            alt_spec_dic.update(alt_spec_var_dic)
+        all_alt_spec_var_dic.update(alt_spec_dic)
+    return all_alt_spec_var_dic
 
-```python
-# # Create a copy of the data subset
-# subset_bike_data = bike_data_long[model_variables].copy()
-```
+##############################################################
+##############################################################
 
-```python
-# # Create the needed variables for the conversion
-# ind_spec_vars = ['num_kids', 'household_size',
-#                  'cars_per_licensed_drivers', 'cross_bay',
-#                  'num_cars', 'num_licensed_drivers']
-
-# alt_spec_vars = ['total_travel_time', 'total_travel_distance']
-
-# subset_spec_vars = {'total_travel_cost': [1, 2, 3],
-#                     'cost_per_distance': [1, 2, 3]}
-
-# alternative_name_dict = {1: 'drive_alone',
-#                          2: 'shared_2',
-#                          3: 'shared_3p',
-#                          4: 'wtw',
-#                          5: 'dtw',
-#                          6: 'wtd',
-#                          7: 'walk',
-#                          8: 'bike'}
-```
+# Code for Trip Specific Variables
+# Loop around the different available alternatives
+def tripSpecDist(data_long, obs_id_col, trip_spec, cont_dists):
+    # Trip Specific Variable (maybe combine with individual specific variables)
```

Comment:
👍 Yes please combine the two functions.

If you truly did have another identifier for trip-id's, then (I assume?) they would follow long-format conventions of not repeating identifying information across columns. In other words, simply passing in a `trip_id_col` would not be sufficient to filter the dataframe to 1 row per decision-maker per trip. You would need the `obs_id_col` and the `trip_id_col`, and you would need to change the logic used to get the data that you want.

As such, this function doesn't do what you would want it to do (get a distribution of values across all unique trips). At the moment, it only duplicates the logic that should be used in `indSpecDist`.

Reply:
Yeah, in general, `tour_id` can be the same for multiple trips on the same tour. However, if I remember correctly, this dataset has been filtered such that there is one tour for going to work/school and one tour returning from work/school. Some exploratory data analysis may answer the question here! Let me know if, when sampling to choice==1, you see individuals with multiple rows of the same `tour_id`. I think you won't due to the pre-processing that occurred.